torch/csrc/distributed/c10d/reducer.cpp

```diff
@@ -951,20 +952,14 @@ void Reducer::finalize_backward() {
 
   // Wait for asynchronous reduction to complete and unflatten contents.
   for (auto& bucket : buckets_) {
-    // See Note [DDP Communication Hook]
-    if (comm_hook_ == nullptr) {
-      TORCH_INTERNAL_ASSERT(
-          bucket.work,
-          "Expected bucket.work not to be null. "
-          "This may indicate that allreduce hooks were not properly installed.");
-      bucket.work->wait();
-    } else {
-      TORCH_INTERNAL_ASSERT(
-          bucket.future_work,
-          "Expected bucket.future_work not to be null. "
-          "This may indicate that communication hook was not properly installed.");
-      bucket.future_work->wait();
+    TORCH_INTERNAL_ASSERT(
+        bucket.future_work,
+        "Expected bucket.future_work not to be null. "
+        "This may indicate that communication hook was not properly installed.");
+    bucket.future_work->wait();
 
+    // See Note [DDP Communication Hook]
+    if (comm_hook_ != nullptr) {
```

Comment:
Instead of checking on `comm_hook` here, we can check if `future_result` is the same as replica.contents and if so, you can skip initializing bucket views since you know it is not a new tensor. You can use the `is_same` API for this: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/templates/TensorBody.h#L194

Reply:
I think that would require `future_value.toTensorVector` first. Instead, I am doing smth like this, please let me know if it isn't good:
https://github.com/pytorch/pytorch/blob/a1056aa822e2e104482526665b0ada27cecbc1ec/torch/csrc/distributed/c10d/reducer.cpp#L957